### ISSUE TITLE
Experimenting autoencoder based approaches for NPM insights

### ISSUE THREAD

**Issue by rootAvish** [openshiftio/openshift.io#2002]
### User Story

As an OSIO/IDE-extensions user I should be able to pass in a stack so that I can get companion/alternate and outlier insights for its components.

### Description

In the [previous experiment done to achieve this target](#1802) it became clear at the start that simply porting the existing model to a different library/framework is not a good approach to tackle large ecosystems, and that with the high availability of data in the NPM ecosystem we can open ourselves to the world of deep learning approaches that generally lead to higher prediction accuracy. We started researching around those models and the research is documented in [this document](https://docs.google.com/document/d/1LTAJRq60lNs-fDGsZLkLS0r8E8j5zDxK_7zkXJqLRxM/edit).

This spike and the [related issue](https://github.com/openshiftio/openshift.io/issues/2004) cover the autoencoder based approaches (CVAE and supervised autoencoder learning). The task list here is incomplete on its own as it has complementing tasks as a part of [related issue](https://github.com/openshiftio/openshift.io/issues/2004).

### Task List

- [ ] ~~Implement the custom layer required for the supervised autoencoder approach~~
- [ ] ~~Implement the custom loss function required for the supervised autoencoder approach~~
- [x] Setup the CVAE code on article recommendation to get a feel for the workflow
- [x] Make changes to the CVAE code so it's able to accommodate our data
- [x] Fit the NPM data in the form of the content matrix and the rating matrix created in [related issue](https://github.com/openshiftio/openshift.io/issues/2004) on the CVAE model.
- [ ] ~~Once done with coding the framework for the supervised autoencoder in [related issue](https://github.com/openshiftio/openshift.io/issues/2004) , put everything together and fit the data to it~~
- [x] Document the findings
---------------------- 
**Optimizations**
- [x] Code the evaluation metrics for the accuracy
- [x] Tuning the learning rate
- [x] Tuning the number of latent factors (hidden layer nodes)
- [ ] T~~uning the momentum factor and dropout regularization~~

Related issue: (https://github.com/openshiftio/openshift.io/issues/2004)
### EPIC: #1809

**Comment by rootAvish** [comment]
Since the CVAE approach worked, and has its results documented here: https://docs.google.com/spreadsheets/d/1OKq1BvPHUzKwYrL8XFWZAgkKC7jcaW0QyY0vleq0wV0/edit?usp=sharing we did not go forward with implementing the supervised autoencoder approach. Work around creating a post-filter business logic for the same will be carried out in this sprint.
